Name: <redacted> <redacted>  
Github Username: <redacted>  
Github Repo: https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>  
Commit ID: 2a1aa470a31ec6c24ba12aa762527853e15a0bcf  

In [1]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

### Load the dataset (load remotely or locally)

In [2]:
iris_data = pd.read_csv("iris.csv")

### Output the first 15 rows of the data

In [3]:
print(f"Iris data (15 rows):\n{iris_data[:15].to_string()}\n")

Iris data (15 rows):
    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14  

### Display a summary of the table information (number of datapoints, etc.)

In [4]:
print(f"Iris data summary:\n{iris_data.describe()}\n")

Iris data summary:
       sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000



## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The data is of sepal and petal lengths and widths of samples of *Iris setosa*, *Iris virginica*, and *Iris versicolor* flowers. There are fifty samples of each of the three species. This (in)famous dataset (as our professor desribes it) was collected by botanist Edgar Anderson, who obtained the samples from "the same pasture, and picked on the same day and measured at the same time by the same person with the same apparatus". The three flowers (all of the genus *Iris*) are notable for their very similar appearance.  
  
The four features are the aforementioned qualities of sepal length, sepal width, petal length, and petal width. The labels are the species of the genus *Iris*, which are either *setosa*, *virginica*, or *versicolor* in this dataset.

# Part 2: Split the dataset into train and test

### Take the dataset and split it into our features (X) and label (y)

In [5]:
iris_X = iris_data.drop("species", axis=1)
iris_y = iris_data["species"]

### Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

In [6]:
X_train, X_test, y_train, y_test = train_test_split(iris_X, iris_y, test_size=0.1, random_state=42)

# Part 3: Logistic Regression

### i. Use sklearn to train a LogisticRegression model on the training set

In [7]:
lr_model = LogisticRegression(max_iter=500)

lr_model.fit(X_train, y_train)

LogisticRegression(max_iter=500)

### ii. For a sample datapoint, predict the probabilities for each possible class

In [8]:
lr_sample_input = X_test.iloc[[1]]
lr_sample_actual_class = y_test.iloc[1]
lr_sample_predicted_class = lr_model.predict(lr_sample_input)[0]

print(f"Input sample point: \n{lr_sample_input}\n")
print(f"Predicted class of sample point: \n{lr_sample_predicted_class}\n")
print(f"Actual class of sample point: \n{lr_sample_actual_class}")

Input sample point: 
    sepal_length  sepal_width  petal_length  petal_width
18           5.7          3.8           1.7          0.3

Predicted class of sample point: 
setosa

Actual class of sample point: 
setosa


### iii. Report on the score for Logistic regression model, what does the score measure?

In [9]:
lr_score = lr_model.score(X_test, y_test)

print(f"Score for LogisticRegression model: {lr_score}")

lr_kf = KFold(n_splits=10, random_state=42, shuffle=True)
lr_cv_scores = cross_val_score(lr_model, X_train, y_train, cv=lr_kf)

print(f"Scores from cross-validation:\n{lr_cv_scores}")
print(f"Average score from cross-validation: {lr_cv_scores.mean()}")

Score for LogisticRegression model: 1.0
Scores from cross-validation:
[1.         1.         1.         0.92857143 1.         0.84615385
 1.         1.         1.         0.92307692]
Average score from cross-validation: 0.9697802197802199


This score measures how accurately the logistic regression model classified the test dataset after being trained on the training set. It does this by reporting what percent of classes had all of its classes identified correctly. Since this dataset only has one class for each data point (it has only one label), this is just the percent of the data it predicted the class of correctly.  
  
It scored a 1.0, which means it predicted the species correctly 100% of the time. This is to be expected for this dataset, which has classes that are quite distinct for the label. Of course, 1.0 prompts some worries of overfitting, so I ran some cross-validation. The mean score from the 10 fold cross-validation was 96.98%, which is very high and alleviates some concern for overfitting while still showing that the model is strong.

### iv. Extract the coefficents and intercepts for the boundary line(s)

In [10]:
lr_classes = lr_model.classes_
lr_coefs = lr_model.coef_
lr_intercept = lr_model.intercept_

print(f"Boundary Lines:")
for i in range(len(lr_classes)):
    c = lr_coefs[i]
    print(f"{lr_classes[i]}: h(x) = {c[0]}x_1 + {c[1]}x_2 + {c[2]}x_3 + {c[3]}x_4 + {lr_intercept[i]}")

Boundary Lines:
setosa: h(x) = -0.4283368821275909x_1 + 0.9673218918184181x_2 + -2.4474144705243437x_3 + -1.03779729472909x_4 + 9.543634962591975
versicolor: h(x) = 0.5130401581400084x_1 + -0.22029920805594355x_2 + -0.21606169206706283x_3 + -0.8445409669682873x_4 + 1.8988147964788542
virginica: h(x) = -0.08470327601242085x_1 + -0.7470226837624787x_2 + 2.663476162591394x_3 + 1.8823382616973798x_4 + -11.442449759071


# Part 4: Support Vector Machine

### i. Use sklearn to train a Support Vector Classifier on the training set

In [11]:
svc_model = svm.SVC(probability=True)

svc_model.fit(X_train, y_train)

SVC(probability=True)

### ii. For a sample datapoint, predict the probabilities for each possible class

In [12]:
svc_classes = svc_model.classes_

svc_sample_input = X_test.iloc[[1]]
svc_sample_actual_class = y_test.iloc[1]
svc_sample_class_probs = svc_model.predict_proba(svc_sample_input)[0]

print(f"Input sample point: \n{svc_sample_input}\n")
print(f"Probabilities of sample point being in each class:")
for i in range(len(svc_sample_class_probs)):
    print(f"{svc_classes[i]}: {svc_sample_class_probs[i] * 100}%")
print(f"\nActual class of sample point: \n{svc_sample_actual_class}")

Input sample point: 
    sepal_length  sepal_width  petal_length  petal_width
18           5.7          3.8           1.7          0.3

Probabilities of sample point being in each class:
setosa: 95.0960638758843%
versicolor: 3.413670731687013%
virginica: 1.490265392428698%

Actual class of sample point: 
setosa


### iii. Report on the score for the SVM, what does the score measure?

In [13]:
svc_score = svc_model.score(X_test, y_test)

print(f"Score for Support Vector Classifier model: {svc_score}")

svc_kf = KFold(n_splits=10, random_state=42, shuffle=True)
svc_scores = cross_val_score(svc_model, X_train, y_train, cv=svc_kf)

print(f"Scores from cross-validation:\n{svc_scores}")
print(f"Average score from cross-validation: {svc_scores.mean()}")

Score for Support Vector Classifier model: 1.0
Scores from cross-validation:
[1.         1.         1.         0.92857143 1.         0.84615385
 1.         0.92307692 1.         1.        ]
Average score from cross-validation: 0.9697802197802197


This score measures how accurately the SVC model classified the test dataset after being trained on the training set. It does this by reporting what percent of classes had all of its classes identified correctly. Since this dataset only has one class for each data point (it has only one label), this is just the percent of the data it predicted the class of correctly.  
  
The model scored a 1.0, which means it predicted the species correctly 100% of the time. As with the Logistic Regression model, this is to be expected for this dataset, which has classes that are quite distinct for the label. Again, 1.0 prompts some worries of overfitting, so I ran some cross-validation. The mean score from the 10 fold cross-validation was 96.98%, which is very high (the same as the Logistic Regression model) and alleviates some concern for overfitting while still showing that the model is strong.

# Part 5: Neural Network

### i. Use sklearn to train a Neural Network (MLP Classifier) on the training set

In [14]:
mlp_model = MLPClassifier(random_state=42, max_iter=600)

mlp_model.fit(X_train, y_train)

MLPClassifier(max_iter=600, random_state=42)

### ii. For a sample datapoint, predict the probabilities for each possible class

In [15]:
mlp_classes = mlp_model.classes_

mlp_sample_input = X_test.iloc[[1]]
mlp_sample_actual_class = y_test.iloc[1]
mlp_sample_class_probs = mlp_model.predict_proba(mlp_sample_input)[0]

print(f"Input sample point: \n{mlp_sample_input}\n")
print(f"Probabilities of sample point being in each class:")
for i in range(len(mlp_sample_class_probs)):
    print(f"{mlp_classes[i]}: {mlp_sample_class_probs[i] * 100}%")
print(f"\nActual class of sample point: \n{mlp_sample_actual_class}")

Input sample point: 
    sepal_length  sepal_width  petal_length  petal_width
18           5.7          3.8           1.7          0.3

Probabilities of sample point being in each class:
setosa: 99.90014019685279%
versicolor: 0.09985980314249929%
virginica: 4.69446990567804e-12%

Actual class of sample point: 
setosa


### iii. Report on the score for the Neural Network, what does the score measure?

In [16]:
mlp_score = mlp_model.score(X_test, y_test)

print(f"Score for Multi-layer Perceptron Classifier model: {mlp_score}")

mlp_kf = KFold(n_splits=10, random_state=42, shuffle=True)
mlp_scores = cross_val_score(mlp_model, X_train, y_train, cv=mlp_kf)

print(f"Scores from cross-validation:\n{mlp_scores}")
print(f"Average score from cross-validation: {mlp_scores.mean()}")

Score for Multi-layer Perceptron Classifier model: 1.0
Scores from cross-validation:
[1.         0.92857143 1.         1.         1.         0.92307692
 1.         0.92307692 1.         0.92307692]
Average score from cross-validation: 0.9697802197802199


This score measures how accurately the MLP classifier model classified the test dataset after being trained on the training set. It does this by reporting what percent of classes had all of its classes identified correctly. Since this dataset only has one class for each data point (it has only one label), this is just the percent of the data it predicted the class of correctly.  
  
Like the previous two models, this model scored a 1.0, which means it predicted the species correctly 100% of the time. As with the other models, this is to be expected for this dataset, which has classes that are quite distinct for the label. 1.0 again prompts some worries of overfitting, so I ran some cross-validation. The mean score from the 10 fold cross-validation was 96.98%, the same as the last two models' cross-validation means, which is a very high score mean, alleviating some concern for overfitting while still showing that the model is strong.

### iv. Experiment with different options for the neural network, report on your best configuration.

My best configuration was the same as my initial testing with 600 max iterations and default settings, which gave a perfect score of 1.0 and a mean score of 96.98% from cross-validation. This may be due to a lucky initial random seed for my test-train split, so I experimented with other seeds and found one that gives worse results for my MLP classifier. With this alternate seed (5 instead of 42), the score became 86.67% and the cross-validation mean score became 98.46%. I found it helpful to raise the maximum iterations to 900, which bumped up the cross-validation mean score to 99.23%. In experimenting I found that the default solver "adam" performed better than the other two (lbfgs and sgd). Smaller batch size reduced cross-validation performance. The logistic sigmoid activation function model performed slightly worse than the identity, hyperbolic tan, and rectified linear activation functions. Other parameters that I tweaked and found no change in performance were hidden layers (tested with (100,), (200, 100, 50), and (100,50,25,10,5)) and learning rate.

# Part 6: K-Nearest Neighbors

### i. Use sklearn to 'train' a k-Neighbors Classifier
Note: KNN is a nonparametric model and technically doesn't require training  
fit will essentially load the data into the model see link below for more information  
https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier

In [17]:
knn_model = KNeighborsClassifier()

knn_model.fit(X_train, y_train)

KNeighborsClassifier()

### ii. For a sample datapoint, predict the probabilities for each possible class

In [18]:
knn_sample_input = X_test.iloc[[1]]
knn_sample_actual_class = y_test.iloc[1]
knn_sample_predicted_class = knn_model.predict(knn_sample_input)[0]

print(f"Input sample point: \n{knn_sample_input}\n")
print(f"Predicted class of sample point: \n{knn_sample_predicted_class}\n")
print(f"Actual class of sample point: \n{knn_sample_actual_class}")

Input sample point: 
    sepal_length  sepal_width  petal_length  petal_width
18           5.7          3.8           1.7          0.3

Predicted class of sample point: 
setosa

Actual class of sample point: 
setosa


### iii. Report on the score for kNN, what does the score measure?

In [19]:
knn_score = knn_model.score(X_test, y_test)

print(f"Score for K-Nearest Neighbors classifier model: {lr_score}")

knn_kf = KFold(n_splits=10, random_state=42, shuffle=True)
knn_cv_scores = cross_val_score(lr_model, X_train, y_train, cv=knn_kf)

print(f"Scores from cross-validation:\n{knn_cv_scores}")
print(f"Average score from cross-validation: {knn_cv_scores.mean()}")

Score for K-Nearest Neighbors classifier model: 1.0
Scores from cross-validation:
[1.         1.         1.         0.92857143 1.         0.84615385
 1.         1.         1.         0.92307692]
Average score from cross-validation: 0.9697802197802199


This score measures how accurately the KNN classifier model classified the test dataset after being trained on the training set. It does this by reporting what percent of classes had all of its classes identified correctly. Since this dataset only has one class for each data point (it has only one label), this is just the percent of the data it predicted the class of correctly.  
Like the previous three models, this model scored a 1.0, which means it predicted the species correctly 100% of the time. As with the other models, this is to be expected for this dataset, which has classes that are quite distinct for the label. Once again, 1.0 prompts some worries of overfitting, so I ran some cross-validation. The mean score from the 10 fold cross-validation was 96.98%, just like all previous models, which is very high and alleviates some concern for overfitting while still showing that the model is strong.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

All models performed equally well on the dataset. All models scored a perfect 1.0 on the training data and cross-validation means were also the same at 0.9698. When checking some other random seeds for the training-testing split, I found that there were some where the models differed slightly. In the seeds I tested, which model performed better varied quite randomly. All models had at least one seed where they did best in cross-validation. The MLP classifier neural network model did best slightly more often, but in one seed it also performed the worst of all of my testing (on all seeds) on the training data.  
  
I think that all of the models performed quite similarly in most cases because the dataset is small and each of the three classes are quite distinct from each other. This allows all models to do very well and very similarly.  
  
The main thing that surprised me about this exercise was exactly what I noted above. I was surprising how well and how similarly all models worked on the data despite taking very different approaches to learning and classifying data.